### ISSUE TITLE
Reduced throughput and tcp retransmissions observed for containers

### ISSUE THREAD

**Issue by nikhildl12** [moby/moby#33115]
**Description**
While doing network performance testing using iperf3, we found that the throughput is reduced also showing a huge number of tcp retransmissions happening when we send traffic from a container as compared to sending traffic from the the host to the same remote iperf server

**Steps to reproduce the issue:**
1. Two hosts with `Ubuntu 16.04.1 LTS (GNU/Linux 4.4.0-57-generic x86_64)`, 10Gbps link
2. Run iperf server on one host: `iperf3 -s`
3. For 1st test, run iperf client on other host: `iperf3 -c 10.158.28.65 -p 5201 -t 30`
4. For 2nd test, run iperf client on other host inside a container: `docker run -ti --rm -p 5201:5201 <path-to-ecr>/iperf:latest iperf3 -c 10.158.28.65 -p 5201 -t 30`

**Describe the results you received:**
```
# Results for host (test 1):
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-30.00  sec  34.6 GBytes  9.90 Gbits/sec    7             sender
[  4]   0.00-30.00  sec  34.6 GBytes  9.90 Gbits/sec                  receiver

# Results for container (test 2):
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-30.00  sec  30.0 GBytes  8.60 Gbits/sec  67701             sender
[  4]   0.00-30.00  sec  30.0 GBytes  8.60 Gbits/sec                  receiver

```
**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   23cf638
 Built:        Tue, 27 Sep 2016 12:25:38 +1300
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   23cf638
 Built:        Tue, 27 Sep 2016 12:25:38 +1300
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 5
 Running: 1
 Paused: 0
 Stopped: 4
Images: 3
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 23
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 4.4.0-57-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 64
Total Memory: 377.8 GiB
Name: slx04c-7vii
ID: 7DHW:WTJP:JIMN:MXZR:IZKG:HMHJ:JJYD:JHK6:HHH5:2OMB:IQMW:LJIU
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8```

**Comment by cpuguy83** [comment]
There is definitely overhead with linux bridges, which most people will never see.
For people that this does effect, there is macvlan and ipvlan.

Also if you are using the port forward, there's the overhead of the NAT...also mitigated by macvlan/ipvlan since you'll have an IP on the host network.

**Comment by cpuguy83** [comment]
Note that you may want to try accessing the container IP directly vs using the NAT to see what the real overhead of the bridge is vs bridge+nat.

**Comment by nikhildl12** [comment]
@cpuguy83: The problem with making the above comparison is that for accessing container IP directly, we have to run iperf server on the same host in a container, and it shows very high throughput since it is on the same host. Where as to test the bridge+NAT case, iperf server has to be running on a different host, which shows the performance mentioned in the test 2 in the description.

Result with accessing container IP directly:
```
docker run -ti --rm iperf:latest iperf3 -c 172.17.0.3 -p 5201
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-20.47  sec  45.9 GBytes  19.3 Gbits/sec    0             sender
[  4]   0.00-20.47  sec  0.00 Bytes  0.00 bits/sec                  receiver
```

**Comment by cpuguy83** [comment]
@nikhildl12 No need to run on the same host, only need to setup static routes to the container's subnet from the other host.

**Comment by nikhildl12** [comment]
@cpuguy83: That's right. So for testing I setup the static routes and I see result similar to using NAT:
```
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-65.79  sec  66.8 GBytes  8.72 Gbits/sec  286601             sender
[  4]   0.00-65.79  sec  0.00 Bytes  0.00 bits/sec                  receiver
```
So I there is not much difference in bridge vs bridge+NAT case

**Comment by cpuguy83** [comment]
@nikhildl12 Did you try macvlan or ipvlan?

**Comment by cpuguy83** [comment]
Also see `--net=host`

**Comment by nikhildl12** [comment]
@cpuguy83: `--net=host` shows same performance as running from host without any issues. I tried `ipvlan` and it does not have the above issue

**Comment by cpuguy83** [comment]
So this overhead is pretty much expected.
Do you think we can close?

**Comment by nikhildl12** [comment]
@cpuguy83: exactly which component should we attribute this overhead to? For testing, I created a test container and attached it to host directly with veth pair. (not to docker0 bridge). And from the host it will be routed to outside. This setup also sees the same issue

**Comment by cpuguy83** [comment]
@nikhildl12 It's the veth/bridge interfaces that have this overhead.

**Comment by nikhildl12** [comment]
@cpuguy83: Another important thing we want to know is even if we agree that the veth/bridge interfaces have overheads which brings down the throughput, are the huge amount tcp retransmissions expected? Are there any such performance tests published comparing the performance hit?

**Comment by cpuguy83** [comment]
There's some comparisons out there, I'm also certain I've seen something more in-depth but can't find it.

http://www.opencloudblog.com/?p=96
https://flavioleitner.blogspot.com/2015/05/open-vswitch-internal-ports-and-linux.html

Mostly comparisons to OVS.
